### DISCUSSION TITLE
Vector Search

### DISCUSSION THREAD

**Discussion by slavakurilyak** [withastro/roadmap#958] [Proposal]
# Summary

Proposal to implement Vector Search capabilities in Astro by leveraging the [native Vector Similarity Search feature of Turso](https://turso.tech/vector), integrated with Astro DB for enhanced AI-powered search functionality.

# Background & Motivation

With the rise of AI-driven applications, there's a growing need for efficient vector search capabilities to support features such as content recommendations, similarity searches, and personalized experiences. Current limitations in Astro's search functionalities hinder developers from fully leveraging AI models for enhanced user experiences. By integrating Turso's native Vector Similarity Search into Astro DB, we can take our search capabilities to the next level, aligning with modern AI trends and developer needs.

# Goals

- **Enhance Search Functionality**: Provide native support for vector search within Astro, allowing developers to implement advanced search features with ease.
- **Improve Developer Experience**: Simplify the process of integrating AI-powered search capabilities into Astro applications.
- **Maintain Performance and Scalability**: Ensure that the new feature is optimized for performance and can scale efficiently with the application's growth.

# Example

The following example demonstrates how to define and query vector data using Astro DB and Drizzle ORM in an Astro application:

### Defining the Schema

```typescript
import { defineDb, defineTable, column } from 'astro:db';

const Movies = defineTable({
  columns: {
    title: column.text({ notNull: true }),
    year: column.number({ notNull: true }),
    embedding: column.floatArray({ notNull: true, length: 32 }),
  },
});

export default defineDb({
  tables: { Movies },
});
```

### Seeding the Database

```typescript
import { Movies } from './db/schema';
import { defineDb } from 'astro:db';

async function seed(db) {
  await db.insert(Movies).values([
    { title: 'Napoleon', year: 2023, embedding: [1, 2, 3] },
    { title: 'Black Hawk Down', year: 2001, embedding: [10, 11, 12] },
    { title: 'Gladiator', year: 2000, embedding: [7, 8, 9] },
    { title: 'Blade Runner', year: 1982, embedding: [4, 5, 6] }
  ]);
}
```

### Querying the Data

To perform a vector search, use the following Drizzle ORM queries:

```typescript
import { Movies } from './db/schema';
import { defineDb } from 'astro:db';

async function searchMovies(db, queryEmbedding) {
  return await db.select({
    title: Movies.columns.title,
    year: Movies.columns.year,
    similarity: db.sql`vector_distance_cos(${Movies.columns.embedding}, ${queryEmbedding}) as similarity`,
  })
  .from(Movies)
  .orderBy(db.sql`similarity`)
  .limit(3);
}
```

### Creating Indexes for Larger Datasets

For larger datasets, implement Approximate Nearest Neighbors (ANN) indexing:

```typescript
import { Movies } from './db/schema';
import { defineDb } from 'astro:db';

async function createIndex(db) {
  await db.run(db.sql`CREATE INDEX movies_idx USING diskann_cosine_ops ON movies (embedding)`);
}

async function searchWithIndex(db, queryEmbedding) {
  return await db.select({
    title: Movies.columns.title,
    year: Movies.columns.year
  })
  .from(db.sql`vector_top_k('movies_idx', ${queryEmbedding}, 3)`)
  .join(Movies, db.sql`${Movies.rowid} = id`)
  .where(db.sql`${Movies.year} >= 2020`);
}
```

By integrating Turso's native vector search capabilities with Astro DB and Drizzle ORM, Astro can significantly enhance its search functionality, providing developers with powerful tools to build next-level, AI-driven applications.

**Comment by zellwk** [comment]
Did a quick search through Astro DB's codebase and Turso's documentation and here's what I found so far: 

1. It looks like we need to add various float types to Astro DB. 

<img width="984" alt="Screenshot 2025-06-21 at 11 50 42 PM" src="https://github.com/user-attachments/assets/528733ad-caa0-4914-bcff-3550f81ef774" />

2. Each of these float based types must include a dimensionality value: 

<img width="1058" alt="Screenshot 2025-06-21 at 11 49 32 PM" src="https://github.com/user-attachments/assets/34c117eb-7c37-4719-a166-a9a312003dee" />

To implement this, we need to: 

1. Include a float schema in the [`schema.ts`](https://github.com/withastro/astro/blob/main/packages/db/src/core/schemas.ts)
2. Add the float type to the [type.ts](https://github.com/withastro/astro/blob/main/packages/db/src/core/types.ts#L39) file. 
3. Add `float` as a type to the  [`schemaTypeToSqlType` ](https://github.com/withastro/astro/blob/main/packages/db/src/core/queries.ts#L100) and return the corresponding sql type.
4. It also looks like we need to implement the relevant conversion function but I'm not sure where this should go: 

<img width="1144" alt="Screenshot 2025-06-21 at 11 50 32 PM" src="https://github.com/user-attachments/assets/c51ef187-15c6-45cf-a6c8-2f1bb0500bd5" />

 ## I'm happy to contribute but I need a bit of help. 

1. I'm not familiar  with `Zod` and Typescript,  so creating the float schema might be a bit tough, especially since I'm not sure how to begin testing the implementation here. 
2. I'm not sure what the corresponding sql type is right now
3. I'm not sure where to add the conversion function at this point

There are a couple of things I'll need to help move this feature along: 

1. We wanna decide on how to implement this — is this in the form @slavakurilyak suggested, or otherwise? 
2. I'll need someone to jump into a 15-30min quick pair programming session with me so I can get up to speed with Zod, Typescript, and the process here and make fewer mistakes in the process.  I should be able to work out the rest afterwards without much problems. 

Thanks for reading this!

**Comment by zellwk** [comment]
I think @slavakurilyak's proposed form is good. But we just need to add the dimensional value within the properties. 

```js
import { defineDb, defineTable, column } from 'astro:db';

const Movies = defineTable({
  columns: {
    title: column.text(),
    year: column.number(),
    embedding: column.floatArray({ length: 32, dimensions: 4 }), // Adds dimensional value 
  },
});

export default defineDb({
  tables: { Movies },
});
```

**Reply by zellwk** [reply]
Gave this a bit more thought. I don't like `floatArray` because it signifies that the type can be used for arrays which is not supported in libsql. That could create confusion. A better name could be `vectorArray` since it gives users a hint that this type would be used for vector embeddings.

**Comment by florian-lefebvre** [comment]
Astro DB has been deprecated
